service/controller/v3/resource/vpngateway/update.go

```diff
@@ -0,0 +1,172 @@
+package vpngateway
+
+import (
+	"context"
+
+	"github.com/Azure/azure-sdk-for-go/services/network/mgmt/2017-09-01/network"
+	providerv1alpha1 "github.com/giantswarm/apiextensions/pkg/apis/provider/v1alpha1"
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/operatorkit/controller"
+
+	"github.com/giantswarm/azure-operator/service/controller/v3/key"
+)
+
+// NewUpdatePatch provide a controller.Patch holding the needed connections update.
+func (r *Resource) NewUpdatePatch(ctx context.Context, azureConfig, current, desired interface{}) (*controller.Patch, error) {
+	a, err := key.ToCustomObject(azureConfig)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	c, err := toVPNGatewayConnections(current)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	d, err := toVPNGatewayConnections(desired)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	patch, err := r.newUpdatePatch(ctx, a, c, d)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	return patch, nil
+}
+
+func (r *Resource) newUpdatePatch(ctx context.Context, azureConfig providerv1alpha1.AzureConfig, current, desired connections) (*controller.Patch, error) {
+	patch := controller.NewPatch()
+
+	change := r.newUpdateChange(ctx, azureConfig, current, desired)
+
+	patch.SetUpdateChange(change)
+
+	return patch, nil
+}
+
+func (r *Resource) newUpdateChange(ctx context.Context, azureConfig providerv1alpha1.AzureConfig, current, desired connections) connections {
+	var change connections
+
+	if needUpdate(current.Host, desired.Host) {
+		change = desired
+	}
+	if needUpdate(current.Guest, desired.Guest) {
+		change = desired
+	}
+
+	return change
+}
+
+// needUpdate determine if current needs to be updated in order to comply with
+// desired. Following properties are examined:
+//
+//     Name
+//     VirtualNetworkGatewayConnectionPropertiesFormat.VirtualNetworkGateway1.ID
+//     VirtualNetworkGatewayConnectionPropertiesFormat.VirtualNetworkGateway2.ID
+//     VirtualNetworkGatewayConnectionPropertiesFormat.ConnectionType
+//     VirtualNetworkGatewayConnectionPropertiesFormat.ConnectionStatus
+//
+func needUpdate(current, desired network.VirtualNetworkGatewayConnection) bool {
```

Comment:
Why do we even use `needUpdate` when we can call `CreateOrUpdate` and it does the heavy lifting for us.

Reply:
The connection shared key is different on every resync period, I am not sure how this will behave when updating the shared key on every resync period. But IMO it is less hassle not to update when not needed.